File: string_manipulate/LC49.py
# 문자열 어레이가 주어졌을 때 아나그램을 묶어라=> 정렬됐을 때 같은 것들을 묶는다
from collections import defaultdict


class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        d = defaultdict(list)
        for word in strs:
            key = "".join(sorted(word))
            d[key].append(word)
        answer = []
        for value in d.values():
            answer.append(value)
        return answer

    # --------------------------------------------------------------------
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        anagrams = defaultdict(list)

        for word in strs:
            anagrams["".join(sorted(word))].append(word)
        return list(anagrams.values())

File: javaalgointerview/LC49.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class LC49 {
	class Solution {
		public List<List<String>> groupAnagrams(String[] strs) {
			Map<String, List<String>> groupMap = new HashMap<>();
			for (String word : strs) {
				char[] c = word.toCharArray();
				Arrays.sort(c);
				String key = Arrays.toString(c);
				groupMap.putIfAbsent(key, new ArrayList<>());
				groupMap.get(key).add(word);
			}
			List<List<String>> answer = new ArrayList<>();
			for (Map.Entry entry : groupMap.entrySet()) {
				answer.add((List<String>)entry.getValue());
			}
			return answer;
		}
	}
}
